classwork/classwork6.1/main.cpp

```diff
@@ -0,0 +1,123 @@
+#include <iostream>
+#include <string>
+#include <functional>
+#include <vector>
+#include <unordered_map>
+#include <thread>
+#include <mutex>
+#include <condition_variable>
+#include <random>
+#include <typeindex>
+#include <atomic>
+#include <print>
+
+struct GreenMessage {
+    std::string text;
+    int counter;
+};
+
+struct BlueMessage {
+    double value1;
+    double value2;
+};
+
+struct OrangeMessage {
+    std::string text1;
+    std::string text2;
+    int counter;
+    double value;
+};
+
+class MessageDispatcher {
+public:
+    template<typename MessageType>
+    void subscribe(std::function<void(const MessageType&)> callback) {
+        std::lock_guard<std::mutex> lock(mtx);
+        auto& subscribers = getSubscribers<MessageType>();
+        subscribers.push_back([callback](const auto& msg) {
+            callback(static_cast<const MessageType&>(msg));
+        });
+    }
+
+    template<typename MessageType>
+    void publish(const MessageType& message) {
+        std::lock_guard<std::mutex> lock(mtx);
+        auto& subscribers = getSubscribers<MessageType>();
+        for (auto& subscriber : subscribers) {
+            subscriber(message);
```

Comment:
it is not a good idea to instantly call all the handlers cause we don't know how long it will take to accomplish them as they are provided by clients. Also in this case real name of the function should be publishAndHandle.
The proper way is to have a separate thread to dispatch messages and invoke callbacks